File: Packages/RallyCore/Sources/RallyCore/Models/AuthModels.swift
import Foundation

/// Response from authentication endpoints.
public struct AuthResponse: Codable, Sendable {
    public let accessToken: String
    public let refreshToken: String
    public let expiresIn: TimeInterval
    public let user: UserProfile

    public init(
        accessToken: String,
        refreshToken: String,
        expiresIn: TimeInterval,
        user: UserProfile
    ) {
        self.accessToken = accessToken
        self.refreshToken = refreshToken
        self.expiresIn = expiresIn
        self.user = user
    }
}

/// Token pair for refresh operations.
public struct TokenPair: Codable, Sendable {
    public let accessToken: String
    public let refreshToken: String
    public let expiresIn: TimeInterval

    public init(accessToken: String, refreshToken: String, expiresIn: TimeInterval) {
        self.accessToken = accessToken
        self.refreshToken = refreshToken
        self.expiresIn = expiresIn
    }
}

/// Request body for Sign in with Apple.
public struct AppleAuthRequest: Codable, Sendable {
    public let identityToken: String
    public let authorizationCode: String
    public let fullName: String?
    public let email: String?

    public init(
        identityToken: String,
        authorizationCode: String,
        fullName: String? = nil,
        email: String? = nil
    ) {
        self.identityToken = identityToken
        self.authorizationCode = authorizationCode
        self.fullName = fullName
        self.email = email
    }
}

/// Result from check-in submission.
public struct CheckInResponse: Codable, Sendable {
    public let checkIn: CheckIn
    public let pointsEarned: Int
    public let newBalance: Int
    public let streakCount: Int
    public let message: String?

    public init(
        checkIn: CheckIn,
        pointsEarned: Int,
        newBalance: Int,
        streakCount: Int = 0,
        message: String? = nil
    ) {
        self.checkIn = checkIn
        self.pointsEarned = pointsEarned
        self.newBalance = newBalance
        self.streakCount = streakCount
        self.message = message
    }
}

/// Result from activation submission (prediction, trivia, etc.).
public struct SubmissionResult: Codable, Sendable {
    public let isCorrect: Bool?
    public let pointsEarned: Int
    public let newBalance: Int
    public let message: String?

    public init(
        isCorrect: Bool? = nil,
        pointsEarned: Int,
        newBalance: Int,
        message: String? = nil
    ) {
        self.isCorrect = isCorrect
        self.pointsEarned = pointsEarned
        self.newBalance = newBalance
        self.message = message
    }
}

/// Result from reward redemption.
public struct RedemptionResult: Codable, Sendable {
    public let redemption: Redemption
    public let newBalance: Int
    public let message: String?

    public init(redemption: Redemption, newBalance: Int, message: String? = nil) {
        self.redemption = redemption
        self.newBalance = newBalance
        self.message = message
    }
}

/// Generic paginated response wrapper.
public struct PaginatedResponse<T: Codable & Sendable>: Codable, Sendable {
    public let items: [T]
    public let page: Int
    public let pageSize: Int
    public let totalItems: Int
    public let totalPages: Int

    public init(items: [T], page: Int, pageSize: Int, totalItems: Int, totalPages: Int) {
        self.items = items
        self.page = page
        self.pageSize = pageSize
        self.totalItems = totalItems
        self.totalPages = totalPages
    }
}

File: Android/core/src/main/java/com/rally/app/core/model/AuthModels.kt
package com.rally.app.core.model

import kotlinx.serialization.Serializable

/**
 * Request body for Google Sign-In authentication.
 */
@Serializable
data class GoogleAuthRequest(
    val idToken: String
)

/**
 * Request body for email/password login.
 */
@Serializable
data class LoginRequest(
    val email: String,
    val password: String,
)

/**
 * Response from email/password login (matches server POST /auth/login).
 */
@Serializable
data class LoginResponse(
    val token: String,
    val user: LoginUser,
)

/**
 * User object returned by the server's /auth/login and /auth/register endpoints.
 */
@Serializable
data class LoginUser(
    val id: String,
    val email: String,
    val name: String,
    val handle: String,
    val role: String = "user",
    val schoolId: String? = null,
    val favoriteSchool: String? = null,
    val points: Int = 0,
    val tier: String = "Bronze",
    val emailVerified: Boolean = false,
)

/**
 * Request body for submitting a poll vote.
 */
@Serializable
data class PollVoteRequest(
    val optionIndex: Int,
)

/**
 * Request body for token refresh.
 */
@Serializable
data class RefreshTokenRequest(
    val refreshToken: String
)

/**
 * Response from authentication endpoints.
 */
@Serializable
data class AuthResponse(
    val accessToken: String,
    val refreshToken: String,
    val expiresIn: Double,
    val user: UserProfile
)

/**
 * Token pair for refresh operations.
 */
@Serializable
data class TokenPair(
    val accessToken: String,
    val refreshToken: String,
    val expiresIn: Double
)

/**
 * Result from check-in submission.
 */
@Serializable
data class CheckInResponse(
    val checkIn: CheckIn,
    val pointsEarned: Int,
    val newBalance: Int,
    val streakCount: Int = 0,
    val message: String? = null
)

/**
 * Result from activation submission (prediction, trivia, etc.).
 */
@Serializable
data class SubmissionResult(
    val isCorrect: Boolean? = null,
    val pointsEarned: Int,
    val newBalance: Int,
    val message: String? = null
)

/**
 * Result from reward redemption.
 */
@Serializable
data class RedemptionResult(
    val redemption: Redemption,
    val newBalance: Int,
    val message: String? = null
)

/**
 * Generic paginated response wrapper.
 */
@Serializable
data class PaginatedResponse<T>(
    val items: List<T>,
    val page: Int,
    val pageSize: Int,
    val totalItems: Int,
    val totalPages: Int
) {
    val hasNextPage: Boolean
        get() = page < totalPages

    val hasPreviousPage: Boolean
        get() = page > 1
}
